File: Neetcode Questions/Trees/binary-tree-level-order-traversal.cpp

// https://leetcode.com/problems/binary-tree-level-order-traversal/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        
        vector<vector<int>> ans;
        if(!root) return ans;

        queue<TreeNode*> q;
        q.push(root);
        q.push(NULL);
        vector<int> smallAns;

        while(!q.empty()){
            auto curr = q.front();
            q.pop();
        
            if(!curr){
                if(q.empty()) break;
                ans.push_back(smallAns);
                smallAns.erase(smallAns.begin(), smallAns.end());
                q.push(NULL);
                continue;
            }

            smallAns.push_back(curr->val);
            if(curr->left) q.push(curr->left);
            if(curr->right) q.push(curr->right);
        }

        ans.push_back(smallAns);
        return ans;
    }
};

File: Neetcode-150-js/Trees/binary-tree-level-order-traversal.js

// https://leetcode.com/problems/binary-tree-level-order-traversal/description/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

Array.prototype.empty = function(){
    return this.length === 0;
}

var levelOrder = function(root) {
    const ans = [];
    if(!root) return ans;

    const q = [root, null];
    let smallAns = [];
    while(!q.empty()){
        const front = q.shift();

        if(!front){
            if(q.empty()) break;
            ans.push(smallAns);
            smallAns = [];
            q.push(null);
            continue;
        }

        smallAns.push(front.val);
        if(front.left) q.push(front.left);
        if(front.right) q.push(front.right);
    }

    ans.push(smallAns);
    return ans;
};